File: server/src/services/dataService.ts
import { BenchmarkType, PricePoint, BacktestResult, BacktestDataPoint, Trade } from "../types";
import { setGlobalDispatcher, ProxyAgent } from 'undici';
import YahooFinance from 'yahoo-finance2';
import axios from 'axios';
import { generateBacktestPythonCode } from './geminiService';

// Set up proxy for undici (fetch)
const proxyUrl = 'http://localhost:4780';
const proxyAgent = new ProxyAgent(proxyUrl);
setGlobalDispatcher(proxyAgent);

// Initialize yahoo-finance2 instance
const yahooFinance = new YahooFinance();

async function fetchYahooFinanceData(ticker: string): Promise<PricePoint[]> {
  try {
    const now = new Date();
    now.setHours(0, 0, 0, 0); // Start of today for stability
    const period1 = new Date(now.getTime() - 365 * 24 * 60 * 60 * 1000); // Exactly 1 year ago from today
    const period2 = now;

    const queryOptions: any = {
      period1,
      period2,
      interval: '1d',
    };

    const result: any = await yahooFinance.chart(ticker, queryOptions);
    
    if (!result || !result.quotes || result.quotes.length === 0) {
      throw new Error(`No data returned for ${ticker}`);
    }

    return (result.quotes as any[])
      .filter(quote => quote.date && quote.close !== null && quote.close !== undefined)
      .map(quote => ({
        date: new Date(quote.date).toISOString().split('T')[0],
        close: quote.close as number,
        open: quote.open as number,
        high: quote.high as number,
        low: quote.low as number,
        volume: quote.volume as number
      }));
  } catch (e: any) {
    console.error(`Yahoo Finance fetch failed for ${ticker}:`, e.message);
    throw e;
  }
}

export const getMarketData = async (benchmark: BenchmarkType): Promise<PricePoint[]> => {
  const tickerMap: Record<string, string> = {
    'BTC-USD': 'BTC-USD',
    'ETH-USD': 'ETH-USD',
    'S&P 500': '^GSPC',
    'CSI 300': '000300.SS'
  };
  
  const ticker = tickerMap[benchmark];
  if (ticker) {
    return await fetchYahooFinanceData(ticker);
  }
  
  throw new Error(`Unsupported benchmark: ${benchmark}`);
};

function buildLongOnlyTrades(data: BacktestDataPoint[], priceSeries: PricePoint[]): Trade[] {
  const priceMap = new Map<string, number>();
  for (const p of priceSeries) {
    if (!priceMap.has(p.date) && typeof p.close === 'number' && p.close > 0) {
      priceMap.set(p.date, p.close);
    }
  }
  const trades: Trade[] = [];
  let position = 0;
  let cash = 100;
  let holdings = 0;
  for (const point of data) {
    const signal = (point as any).signal as 'BUY' | 'SELL' | undefined;
    if (!signal) continue;
    const price = priceMap.get(point.date);
    if (!price || price <= 0) continue;
    if (signal === 'BUY' && position === 0) {
      const quantity = cash / price;
      const amount = quantity * price;
      trades.push({
        date: point.date,
        type: 'BUY',
        price,
        quantity,
        amount
      });
      holdings = quantity;
      cash = 0;
      position = 1;
    } else if (signal === 'SELL' && position === 1) {
      const amount = holdings * price;
      trades.push({
        date: point.date,
        type: 'SELL',
        price,
        quantity: holdings,
        amount
      });
      cash = amount;
      holdings = 0;
      position = 0;
    }
  }
  if (position === 1 && data.length > 0) {
    const lastPoint = data[data.length - 1];
    const price = priceMap.get(lastPoint.date);
    if (price && price > 0 && holdings > 0) {
      const amount = holdings * price;
      trades.push({
        date: lastPoint.date,
        type: 'SELL',
        price,
        quantity: holdings,
        amount
      });
    }
  }
  return trades;
}

export const runBacktest = async (
  formula: string, 
  benchmark: BenchmarkType,
  buyThreshold?: string,
  sellThreshold?: string,
  requestId?: string,
  pythonCode?: string
): Promise<BacktestResult> => {
  const id = requestId || `bk-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;
  console.log(`[Backtest] [${id}] Starting backtest benchmark=${benchmark}`);
  console.log(`[Backtest] [${id}] Formula length=${formula.length} buy=${buyThreshold || '-'} sell=${sellThreshold || '-'}`);

  const priceData = await getMarketData(benchmark);
  console.log(`[Backtest] [${id}] Loaded market data points=${priceData.length}`);
  
  try {
    const pythonScript = pythonCode || await generateBacktestPythonCode(formula);
    console.log(`[Backtest] [${id}] Generated python script : \n ${pythonScript}`);
    
    try {
      const response = await axios.post('http://localhost:5001/execute', {
        code: pythonScript,
        data: {
          priceData,
          formula,
          benchmark,
          buyThreshold,
          sellThreshold
        }
      });

      const result = response.data;
      console.log(`[Backtest] [${id}] Python service status=${result.status}`);
      
      if (result.status === 'error') {
        console.error(`[Backtest] [${id}] Python error=${result.error}`);
        throw new Error(`Python Service Error: ${result.error}\nStdout: ${result.stdout}`);
      }
      
      if (result.result && result.result.metrics) {
        const m = result.result.metrics;
        console.log(`[Backtest] [${id}] Metrics sharpe=${m.sharpeRatio} annReturn=${m.annualizedReturn} maxDD=${m.maxDrawdown}`);
      }

      let payload: BacktestResult = {
        ...result.result,
        pythonCode: pythonScript
      };
      if (Array.isArray(payload?.data) && payload.data.length > 0) {
        const syntheticTrades = buildLongOnlyTrades(payload.data, priceData);
        if (syntheticTrades.length > 0) {
          payload = {
            ...payload,
            trades: syntheticTrades
          };
        }
      }

      console.log(`[Backtest] [${id}] Completed successfully with ${payload.trades?.length || 0} trades`);
      return payload;

    } catch (e: any) {
      if (axios.isAxiosError(e)) {
        console.error(`[Backtest] [${id}] Python HTTP error=${e.message}`);
        throw new Error(`Python Service HTTP Error: ${e.message} - ${JSON.stringify(e.response?.data)}`);
      }
      console.error(`[Backtest] [${id}] Python execution error=${e.message}`);
      throw new Error(`Failed to execute backtest on python service: ${e.message}`);
    }

  } catch (e: any) {
    console.error(`[Backtest] [${id}] Backtest failed=${e.message}`);
    throw new Error(`Failed to run backtest: ${e.message}`);
  }
};

File: server_py/data_service.py
from typing import List, Mapping, Optional

import os
import re
import requests
import yfinance as yf

from .executor import execute_backtest
from .gemini_service import generate_backtest_python_code
from .models import (
    BacktestDataPoint,
    BacktestResult,
    BenchmarkType,
    PricePoint,
    Trade,
)

def _fetch_yahoo_finance_data(ticker: str) -> List[PricePoint]:
    df = yf.download(
        ticker,
        period="1y",
        interval="1d",
        auto_adjust=False,
        progress=False,
    )
    if df.empty:
        raise RuntimeError(f"No data returned for {ticker}")
    records: List[PricePoint] = []
    for index, row in df.iterrows():
        close = float(row.get("Close"))
        if not close:
            continue
        price_point = PricePoint(
            date=index.strftime("%Y-%m-%d"),
            close=close,
            open=float(row.get("Open")) if row.get("Open") is not None else None,
            high=float(row.get("High")) if row.get("High") is not None else None,
            low=float(row.get("Low")) if row.get("Low") is not None else None,
            volume=float(row.get("Volume")) if row.get("Volume") is not None else None,
        )
        records.append(price_point)
    if not records:
        raise RuntimeError(f"No valid price points for {ticker}")
    return records


def get_market_data(benchmark: BenchmarkType) -> List[PricePoint]:
    ticker_map: Mapping[BenchmarkType, str] = {
        BenchmarkType.BTC_USD: "BTC-USD",
        BenchmarkType.ETH_USD: "ETH-USD",
        BenchmarkType.SP500: "^GSPC",
        BenchmarkType.CSI300: "000300.SS",
    }
    ticker = ticker_map.get(benchmark)
    if not ticker:
        raise ValueError(f"Unsupported benchmark: {benchmark}")
    return _fetch_yahoo_finance_data(ticker)


def _normalize_a_share_ticker(code: str) -> str:
    raw = (code or "").strip()
    if not raw or not re.fullmatch(r"\d{6}", raw):
        raise ValueError(f"Invalid A-share code: {code}")
    first = raw[0]
    if first in ("6", "9"):
        suffix = ".SS"
    else:
        suffix = ".SZ"
    return f"{raw}{suffix}"


def get_a_share_market_data_from_code(code: str) -> List[PricePoint]:
    ticker = _normalize_a_share_ticker(code)
    return _fetch_yahoo_finance_data(ticker)


def _build_long_only_trades(
    data: List[BacktestDataPoint], price_series: List[PricePoint]
) -> List[Trade]:
    price_map: dict[str, float] = {}
    for p in price_series:
        if p.date not in price_map and p.close and p.close > 0:
            price_map[p.date] = p.close
    trades: List[Trade] = []
    position = 0
    cash = 100.0
    holdings = 0.0
    for point in data:
        signal = point.signal
        if not signal:
            continue
        price = price_map.get(point.date)
        if not price or price <= 0:
            continue
        if signal == "BUY" and position == 0:
            quantity = cash / price
            amount = quantity * price
            trades.append(
                Trade(
                    date=point.date,
                    type="BUY",
                    price=price,
                    quantity=quantity,
                    amount=amount,
                )
            )
            holdings = quantity
            cash = 0.0
            position = 1
        elif signal == "SELL" and position == 1:
            amount = holdings * price
            trades.append(
                Trade(
                    date=point.date,
                    type="SELL",
                    price=price,
                    quantity=holdings,
                    amount=amount,
                )
            )
            cash = amount
            holdings = 0.0
            position = 0
    if position == 1 and data:
        last_point = data[-1]
        price = price_map.get(last_point.date)
        if price and price > 0 and holdings > 0:
            amount = holdings * price
            trades.append(
                Trade(
                    date=last_point.date,
                    type="SELL",
                    price=price,
                    quantity=holdings,
                    amount=amount,
                )
            )
    return trades


def run_backtest(
    formula: str,
    benchmark: BenchmarkType,
    buyThreshold: Optional[str] = None,
    sellThreshold: Optional[str] = None,
    pythonCode: Optional[str] = None,
    customCode: Optional[str] = None,
) -> BacktestResult:
    import time

    request_id = f"bk-{int(time.time() * 1000):x}"
    print(f"[Backtest] [{request_id}] Starting backtest benchmark={benchmark}")
    print(
        f"[Backtest] [{request_id}] Formula length={len(formula)} buy={buyThreshold or '-'} sell={sellThreshold or '-'}"
    )
    if customCode:
        print(f"[Backtest] [{request_id}] Using custom A-share code={customCode}")
        price_data = get_a_share_market_data_from_code(customCode)
        benchmark_label = customCode
    else:
        price_data = get_market_data(benchmark)
        benchmark_label = benchmark.value
    print(f"[Backtest] [{request_id}] Loaded market data points={len(price_data)}")
    try:
        python_script = pythonCode or generate_backtest_python_code(formula)
        print(f"[Backtest] [{request_id}] Generated python script:\n{python_script}")
        payload = execute_backtest(
            python_script,
            {
                "priceData": [p.dict() for p in price_data],
                "formula": formula,
                "benchmark": benchmark_label,
                "buyThreshold": buyThreshold,
                "sellThreshold": sellThreshold,
            },
        )
        status = payload.get("status")
        print(f"[Backtest] [{request_id}] Python service status={status}")
        if status == "error":
            error_message = payload.get("error") or "Unknown error"
            stdout = payload.get("stdout") or ""
            print(f"[Backtest] [{request_id}] Python error={error_message}")
            raise RuntimeError(
                f"Python Service Error: {error_message}\nStdout: {stdout}"
            )
        result_data = payload.get("result") or {}
        if result_data.get("metrics"):
            m = result_data["metrics"]
            print(
                f"[Backtest] [{request_id}] Metrics sharpe={m.get('sharpeRatio')} annReturn={m.get('annualizedReturn')} maxDD={m.get('maxDrawdown')}"
            )
        result = BacktestResult(**result_data)
        if result.data:
            synthetic_trades = _build_long_only_trades(result.data, price_data)
            if synthetic_trades:
                result.trades = synthetic_trades
        print(
            f"[Backtest] [{request_id}] Completed successfully with {len(result.trades)} trades"
        )
        result.pythonCode = python_script
        return result
    except Exception as exc:
        print(f"[Backtest] [{request_id}] Backtest failed={exc}")
        raise RuntimeError(f"Failed to run backtest: {exc}") from exc
